processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +160,44 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
-        throw new IndexSizeExceededException("Maximum number of rows reached");
-      }
-      final Integer prev = facts.putIfAbsent(key, rowOffset);
-      if (prev != null) {
-        rowOffset = prev;
-      } else {
-        Aggregator[] aggs = new Aggregator[metrics.length];
-        for (int i = 0; i < metrics.length; i++) {
-          final AggregatorFactory agg = metrics[i];
-          aggs[i] = agg.factorize(
-              makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
-          );
-        }
-        aggList.add(aggs);
-        numEntries.incrementAndGet();
+    final Aggregator[] aggs;
+    final Integer retval;
+    final Integer rowOffset = numEntries.get();
+    if (rowOffset >= maxRowCount && !facts.containsKey(key)) {
+      throw new IndexSizeExceededException("Maximum number of rows reached");
+    }
+    final Integer prev = facts.putIfAbsent(key, rowOffset);
+    if (prev != null) {
+      aggs = aggList.get(prev);
+      retval = rowOffset;
+    } else {
+      aggs = new Aggregator[metrics.length];
+      for (int i = 0; i < metrics.length; i++) {
+        final AggregatorFactory agg = metrics[i];
+        aggs[i] = agg.factorize(
+            makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
+        );
       }
+      aggList.set(rowOffset, aggs);
+      retval = numEntries.incrementAndGet();
     }
 
     in.set(row);
 
-    final Aggregator[] aggs = aggList.get(rowOffset);
     for (int i = 0; i < aggs.length; i++) {
       synchronized (aggs[i]) {
         aggs[i].aggregate();
       }
     }
     in.set(null);
-    return numEntries.get();
+    return retval;
```

Comment:
This changes the current behavior in a concurrent setting. Not sure if this affects anything, but before we would return the number of rows in the index at the time the function exits, whereas now we return the row number for the row just added.